js/ui/paygUnlockDialog.js

```diff
@@ -0,0 +1,395 @@
+                                             reactive: true,
+                                             can_focus: true,
+                                             label: _("Cancel"),
+                                             x_align: St.Align.START,
+                                             y_align: St.Align.END });
+        buttonsBox.add_child(this._cancelButton);
+
+        let buttonSpacer = new St.Widget({ layout_manager: new Clutter.BinLayout(),
+                                             x_expand: true,
+                                             x_align: Clutter.ActorAlign.END });
+        buttonsBox.add_child(buttonSpacer);
+
+        // We make the most of the spacer to show the spinner while verifying the code.
+        let spinnerIcon = Gio.File.new_for_uri('resource:///org/gnome/shell/theme/process-working.svg');
+        this._spinner = new Animation.AnimatedIcon(spinnerIcon, SPINNER_ICON_SIZE);
+        this._spinner.actor.opacity = 0;
+        this._spinner.actor.show();
+        buttonSpacer.add_child(this._spinner.actor);
+
+        this._nextButton = new St.Button({ style_class: 'modal-dialog-button button',
+                                           button_mask: St.ButtonMask.ONE | St.ButtonMask.THREE,
+                                           reactive: true,
+                                           can_focus: true,
+                                           label: _("Unlock"),
+                                           x_align: St.Align.END,
+                                           y_align: St.Align.END });
+        this._nextButton.add_style_pseudo_class('default');
+        buttonsBox.add_child(this._nextButton);
+
+        return buttonsBox;
+    },
+
+    _onCancelled: function() {
+        this._cancelled = true;
+        this._reset();
+
+        // The ScreenShield will connect to the 'failed' signal
+        // to know when to cancel the unlock dialog.
+        if (this._verificationStatus != UnlockStatus.SUCCEEDED)
+            this.emit('failed');
+    },
+
+    _validateCurrentCode: function() {
+        // The PaygUnlockCodeEntry widget will only accept valid
+        // characters, so we only need to check the length here.
+        return this._entry.length == CODE_ENTRY_REQUIRED_LENGTH;
+    },
+
+    _updateNextButtonSensitivity: function() {
+        let sensitive = this._validateCurrentCode() && this._verificationStatus != UnlockStatus.VERIFYING;
+        this._nextButton.reactive = sensitive;
+        this._nextButton.can_focus = sensitive;
+    },
+
+    _updateSensitivity: function() {
+        this._updateNextButtonSensitivity();
+        this._entry.setEnabled(this._verificationStatus != UnlockStatus.VERIFYING);
+    },
+
+    _setErrorMessage: function(message) {
+        if (message) {
+            this._errorMessage.text = message;
+            this._errorMessage.opacity = 255;
```

Comment:
I assume you’re hiding/showing `this._errorMessage` using opacity rather than `.show()`/`.hide()`, to avoid it causing relayouts?

Reply:
Actually, I'm not 100% sure, since I'm inheriting this "technique" from the unlockDialog.js module, although I thought exactly the same at the time.